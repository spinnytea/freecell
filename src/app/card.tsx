import Image from "next/image";

// TODO red & black joker
export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];

export type Rank = 'ace' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | 'jack' | 'queen' | 'king';
export const RankList: Rank[] = ['ace', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'jack', 'queen', 'king'];

const ORIG_WIDTH = 167.0869141;
const ORIG_HEIGHT = 242.6669922;
const scale_height = (width: number) => Math.floor(width / ORIG_WIDTH * ORIG_HEIGHT);

export function CardImage({
	width=ORIG_WIDTH,
	rank,
	suit,
}: {
	width?: number,
	rank: Rank,
	suit: Suit,
}) {
	const filename = getFilename(rank, suit);
	const height = scale_height(width);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={width}
			height={height}
			priority
		/>
	);
}

function getFilename(rank: Rank, suit: Suit) {
	const fancy = (rank === 'jack' || rank === 'queen' || rank === 'king' ? '2' : '');
	return `/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}
