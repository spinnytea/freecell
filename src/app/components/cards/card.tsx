import Image from 'next/image';

// TODO red & black joker
export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];

const ORIG_WIDTH = 167.0869141;
const ORIG_HEIGHT = 242.6669922;
const scale_height = (width: number) => Math.floor((width / ORIG_WIDTH) * ORIG_HEIGHT);

const FANCY_DECK = true;

export function CardImage({
	width = ORIG_WIDTH,
	rank,
	suit,
}: {
	width?: number;
	rank: Rank;
	suit: Suit;
}) {
	// REVIEW do we need `priority` - it's complaining because i'm rendering the whole deck on init
	const filename = getFilename(rank, suit);
	const height = scale_height(width);
	return (
		<Image
			src={filename}
			alt={`${rank} of ${suit}`}
			width={Math.floor(width)}
			height={height}
			priority
		/>
	);
}

// TODO toggle fancy (also ace of spades)
function getFilename(rank: Rank, suit: Suit, useFancyDeck = FANCY_DECK) {
	if (rank === 'joker') {
		if (isRed(suit)) return `/SVG-cards-1.3/red_joker.svg`;
		return `/SVG-cards-1.3/black_joker.svg`;
	}
	if (rank === 'ace' && suit === 'spades') {
		if (useFancyDeck) return `/SVG-cards-1.3/ace_of_spades.svg`;
		return `/SVG-cards-1.3/ace_of_spades2.svg`;
	}
	const fancy = useFancyDeck
		? rank === 'jack' || rank === 'queen' || rank === 'king'
			? '2'
			: ''
		: '';
	return `/SVG-cards-1.3/${rank}_of_${suit}${fancy}.svg`;
}

/** stuff exported for testing, not meant to be used directly */
export const units = {
	getFilename,
};
