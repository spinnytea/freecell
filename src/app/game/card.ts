// XXX (techdebt) duplicate
const BOTTOM_OF_CASCADE = 99;

export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
export const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];
export const isAdjacent = ({ min, max }: { min: Rank; max: Rank }) =>
	RankList.indexOf(min) === RankList.indexOf(max) - 1;

// REVIEW (techdebt) is "fixture" the right name?
//  - cell -> freecell
//  - foundation -> homecell
//  - cascade -> column
export type Fixture = 'deck' | 'cell' | 'foundation' | 'cascade';
export interface CardLocation {
	fixture: Fixture;
	data: number[];
}

export type MoveSourceType = 'deck' | 'cell' | 'foundation' | 'cascade:single' | 'cascade:sequence';
export type MoveDestinationType = 'cell' | 'foundation' | 'cascade:empty' | 'cascade:sequence';
export const MoveDestinationTypePriorities: {
	[moveSourceType in MoveSourceType]: { [moveDestinationType in MoveDestinationType]: number };
} = {
	// XXX (controls) deck: when will we get to do this?
	'deck': {
		'cell': 1,
		'foundation': 4,
		'cascade:empty': 2,
		'cascade:sequence': 3,
	},
	'cell': {
		'cell': 1,
		'foundation': 3,
		'cascade:empty': 2,
		'cascade:sequence': 4,
	},
	// XXX (controls) foundation: down from foundation means "back into play?"
	'foundation': {
		'cell': 1,
		'foundation': 4,
		'cascade:empty': 2,
		'cascade:sequence': 3,
	},
	'cascade:single': {
		'cell': 2,
		'foundation': 3,
		'cascade:empty': 1,
		'cascade:sequence': 4,
	},
	'cascade:sequence': {
		'cell': 1,
		'foundation': 2,
		'cascade:empty': 3,
		'cascade:sequence': 4,
	},
};

/**
	REVIEW (techdebt) does AvailableMove need to store moveDestinationType and priority?
	 - moveDestinationType is needed for building the list (like moveSourceType)
	   but do we need to store it?
	 - priority could be build into the sort order of the list (pick the first item),
	   rather than stored with the location (pick the highest number)
	 - it is nice to see in the DebugCursors the different categories
	 	priority: 'high' | 'low';
	---
	 - we have sufficient impl to simplify now
	 - wait for unit testing of DebugCursors
	 - wait for manualtesting of DebugCursors
*/
export interface AvailableMove {
	/** where we could move the card */
	location: CardLocation;
	/** helps us think about priorities / communicate settings */
	moveDestinationType: MoveDestinationType;
	/** if we are going to visualize them debug mode, we need to have it precomputed */
	priority: number;
}

/**
	foundation: h
	cells: a - d
	cascades: 1 - 9, t (1-8, but we allow 9 and 10 columns)

	@see [Standard FreeCell Notation](https://www.solitairelaboratory.com/solutioncatalog.html)
*/
export type Position =
	| 'a'
	| 'b'
	| 'c'
	| 'd'
	| 'h'
	| '1'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| 't';

export interface Card {
	rank: Rank;
	suit: Suit;
	location: CardLocation;
}

export interface CardSequence {
	/**
		the first card in the sequence
		(cursor, selection)
	*/
	location: CardLocation;

	/**
		the list of cards in the sequence
		(alternates red/black, rank descends by one)
	*/
	cards: Card[];

	/** since we are allowing any card to be selected or inspected, not all of them are movable */
	canMove: boolean;
}

export function isLocationEqual(a: CardLocation, b: CardLocation) {
	return a.fixture === b.fixture && a.data[0] === b.data[0] && a.data[1] === b.data[1];
}

export function shorthandCard(card: Card | null | undefined) {
	if (!card) return '  ';
	const r = card.rank === '10' ? 'T' : card.rank === 'joker' ? 'W' : card.rank[0];
	const s = card.suit[0];
	return (r + s).toUpperCase();
}

export function parseShorthandCard(
	r: string | undefined,
	s: string | undefined
): { rank: Rank; suit: Suit } | null {
	if (r === ' ' && s === ' ') return null;

	let rank: Rank;
	let suit: Suit;
	switch (r) {
		case 'A':
			rank = 'ace';
			break;
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			rank = r;
			break;
		case 'T':
			rank = '10';
			break;
		case 'J':
			rank = 'jack';
			break;
		case 'Q':
			rank = 'queen';
			break;
		case 'K':
			rank = 'king';
			break;
		case 'W':
			rank = 'joker';
			break;
		default:
			throw new Error(`invalid rank shorthand: "${r ?? 'undefined'}"`);
	}

	switch (s) {
		case 'C':
			suit = 'clubs';
			break;
		case 'D':
			suit = 'diamonds';
			break;
		case 'H':
			suit = 'hearts';
			break;
		case 'S':
			suit = 'spades';
			break;
		default:
			throw new Error(`invalid suit shorthand: "${s ?? 'undefined'}"`);
	}

	return { rank, suit };
}

export function shorthandSequence(sequence: CardSequence, includePosition = false) {
	const cards = sequence.cards.map((card) => shorthandCard(card)).join('-');

	if (sequence.canMove && includePosition) {
		return shorthandPosition(sequence.location) + ' ' + cards;
	}

	return cards;
}

export function shorthandPosition(location: CardLocation): Position {
	const d0 = location.data[0];
	if (location.fixture === 'foundation') {
		return 'h';
	} else if (location.fixture === 'cascade') {
		// this doesn't check data[1], it assumes it's the final row
		// sequences would need to check data[1] + card.length, not just the location
		// (could pass in a optional canMove with default of true)
		if (d0 === 9) {
			return 't';
		} else if (d0 >= 0 && d0 < 9) {
			return (d0 + 1).toString(10) as Position;
		}
	} else if (location.fixture === 'cell') {
		if (d0 >= 0 && d0 < 6) {
			return (d0 + 10).toString(16) as Position;
		}
	}
	throw new Error(`invalid position: ${JSON.stringify(location)}`);
}

/**
	this is part 1 of a 2 step process
	moves are not always obvious
	i.e. 1h - _which_ foundation do we use? there are 4
	i.e. 42 - is this a cascade:single or cascade:sequence moving to a cascade:sequence or cascade:empty?

	notice also that this function only accepts the single character, it does not accept a game
	so which d1 do we use for a cascade? this will return an invalid value (too high), which will be clamped if used directly
*/
export function parseShorthandPosition_INCOMPLETE(p: string | undefined): CardLocation {
	switch (p) {
		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			// this isn't a valid cursor position, it will need to be clamped
			// cascades can have sequences, so you need to decide if you really want the "bottom"
			return { fixture: 'cascade', data: [parseInt(p, 10) - 1, BOTTOM_OF_CASCADE] };
		case 't':
			return { fixture: 'cascade', data: [9, BOTTOM_OF_CASCADE] };
		case 'h':
			// h could refer to _any_ of the foundations; this needs to be verified
			return { fixture: 'foundation', data: [0] };
		case 'a':
		case 'b':
		case 'c':
		case 'd':
		case 'e':
		case 'f':
			return { fixture: 'cell', data: [parseInt(p, 16) - 10] };
		default:
			throw new Error(`invalid position shorthand: "${p ?? 'undefined'}"`);
	}
}
