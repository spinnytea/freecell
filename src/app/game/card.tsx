export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
export const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];

export type Fixture = 'deck' | 'cell' | 'foundation' | 'cascade';
export interface CardLocation {
	fixture: Fixture;
	data: number[];
}

export interface Card {
	rank: Rank;
	suit: Suit;
	location: CardLocation;
}

export function shorthand(card: Card | null | undefined) {
	if (!card) return '  ';
	const r = card.rank === '10' ? 'T' : card.rank[0];
	const s = card.suit[0];
	return (r + s).toUpperCase();
}
