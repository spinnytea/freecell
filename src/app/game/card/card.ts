import { BOTTOM_OF_CASCADE } from '@/app/components/cards/constants';
import { FreeCell } from '@/app/game/game';

/* *********** */
/* DEFINITIONS */
/* *********** */

export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
export const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];
export const getRankForCompare = (rank: Rank): number => RankList.indexOf(rank);
export const isAdjacent = ({ min, max }: { min: Rank; max: Rank }) =>
	getRankForCompare(min) === getRankForCompare(max) - 1;

// XXX (techdebt) is "fixture" the right name?
//  - cell -> freecell
//  - foundation -> homecell
//  - cascade -> column
export type Fixture = 'deck' | 'cell' | 'foundation' | 'cascade';
export interface CardLocation {
	fixture: Fixture;
	data: number[];
}

/**
	foundation: h
	cells: a - d
	cascades: 1 - 9, t (1-8, but we allow 9 and 10 columns)

	@see [Standard FreeCell Notation](https://www.solitairelaboratory.com/solutioncatalog.html)
*/
export type Position =
	| 'a'
	| 'b'
	| 'c'
	| 'd'
	| 'e'
	| 'f'
	| 'h'
	| '1'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '0';

export interface CardSH {
	rank: Rank;
	suit: Suit;
}
export interface Card extends CardSH {
	location: CardLocation;
}

export interface CardSequence {
	/**
		the first card in the sequence
		(cursor, selection)
	*/
	location: CardLocation;

	/**
		the list of cards in the sequence
		(alternates red/black, rank descends by one)
	*/
	cards: Card[];

	/** since we are allowing any card to be selected or inspected, not all of them are movable */
	canMove: boolean;
}

/* ************** */
/* HELPER METHODS */
/* ************** */

export function cloneCards(cards: Card[]) {
	// cards need to remain in consitent order for react[key=""] to work
	return cards.map((card) => ({ ...card }));
}

export function findCard(cards: Card[], card: CardSH | null | undefined): Card {
	if (!card) throw new Error('no card provided');
	const found = cards.find((c) => c.suit === card.suit && c.rank === card.rank);
	if (!found) throw new Error('missing card ' + shorthandCard(card));
	return found;
}

export function isLocationEqual(a: CardLocation, b: CardLocation) {
	return a.fixture === b.fixture && a.data[0] === b.data[0] && a.data[1] === b.data[1];
}

export function getSequenceAt(game: FreeCell, location: CardLocation): CardSequence {
	const [d0] = location.data;

	switch (location.fixture) {
		case 'deck':
			{
				const card = game.deck[d0];
				// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
				if (card) {
					return {
						location,
						cards: [card],
						canMove: false,
					};
				}
			}
			break;
		case 'foundation':
			{
				const card = game.foundations[d0];
				if (card) {
					return {
						location,
						cards: [card],
						canMove: false,
					};
				}
			}
			break;
		case 'cell':
			{
				const card = game.cells[d0];
				if (card) {
					return {
						location,
						cards: [card],
						canMove: true,
					};
				}
			}
			break;
		case 'cascade': {
			const cascade = game.tableau[d0];
			let idx = location.data[1];

			if (!cascade[idx]) break;

			const sequence: CardSequence = {
				location,
				cards: [cascade[idx]],
				canMove: false,
			};

			while (
				idx < cascade.length - 1 &&
				isAdjacent({ min: cascade[idx + 1].rank, max: cascade[idx].rank }) &&
				isRed(cascade[idx].suit) !== isRed(cascade[idx + 1].suit)
			) {
				idx++;
				sequence.cards.push(cascade[idx]);
			}

			if (idx === cascade.length - 1) {
				sequence.canMove = true;
			}

			return sequence;
		}
	}

	// no cards at selection
	return { location, cards: [], canMove: false };
}

/* ************* */
/* PRINT / PARSE */
/* ************* */

export function shorthandCard(card: CardSH | null | undefined) {
	if (!card) return '  ';
	const r = card.rank === '10' ? 'T' : card.rank === 'joker' ? 'W' : card.rank[0];
	const s = card.suit[0];
	return (r + s).toUpperCase();
}

/** TODO (techdebt) (refactor) change `rs` to single string since that's how it's _always_ called */
export function parseShorthandCard(r: string | undefined, s: string | undefined): CardSH | null {
	if (r === ' ' && s === ' ') return null;

	let rank: Rank;
	let suit: Suit;
	switch (r) {
		case 'A':
			rank = 'ace';
			break;
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			rank = r;
			break;
		case 'T':
			rank = '10';
			break;
		case 'J':
			rank = 'jack';
			break;
		case 'Q':
			rank = 'queen';
			break;
		case 'K':
			rank = 'king';
			break;
		case 'W':
			rank = 'joker';
			break;
		default:
			throw new Error(`invalid rank shorthand: "${r ?? 'undefined'}"`);
	}

	switch (s) {
		case 'C':
			suit = 'clubs';
			break;
		case 'D':
			suit = 'diamonds';
			break;
		case 'H':
			suit = 'hearts';
			break;
		case 'S':
			suit = 'spades';
			break;
		default:
			throw new Error(`invalid suit shorthand: "${s ?? 'undefined'}"`);
	}

	return { rank, suit };
}

export function shorthandSequence(sequence: CardSequence, includePosition = false) {
	const cards = sequence.cards.map((card) => shorthandCard(card)).join('-');

	if (sequence.canMove && includePosition) {
		return shorthandPosition(sequence.location) + ' ' + cards;
	}

	return cards;
}

export function shorthandPosition(location: CardLocation): Position {
	const d0 = location.data[0];
	if (location.fixture === 'foundation') {
		return 'h';
	} else if (location.fixture === 'cascade') {
		// this doesn't check data[1], it assumes it's the final row
		// sequences would need to check data[1] + card.length, not just the location
		// (could pass in a optional canMove with default of true)
		if (d0 === 9) {
			return '0';
		} else if (d0 >= 0 && d0 < 9) {
			return (d0 + 1).toString(10) as Position;
		}
	} else if (location.fixture === 'cell') {
		if (d0 >= 0 && d0 < 6) {
			return (d0 + 10).toString(16) as Position;
		}
	}
	throw new Error(`invalid position: ${JSON.stringify(location)}`);
}

/**
	this is part 1 of a 2 step process
	moves are not always obvious
	i.e. 1h - _which_ foundation do we use? there are 4
	i.e. 42 - is this a cascade:single or cascade:sequence moving to a cascade:sequence or cascade:empty?

	notice also that this function only accepts the single character, it does not accept a game
	so which d1 do we use for a cascade? this will return an invalid value (too high), which will be clamped if used directly
*/
export function parseShorthandPosition_INCOMPLETE(p: string | undefined): CardLocation {
	switch (p) {
		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			// this isn't a valid cursor position, it will need to be clamped
			// cascades can have sequences, so you need to decide if you really want the "bottom"
			return { fixture: 'cascade', data: [parseInt(p, 10) - 1, BOTTOM_OF_CASCADE] };
		// ten
		case '0':
			return { fixture: 'cascade', data: [9, BOTTOM_OF_CASCADE] };
		case 'h':
			// h could refer to _any_ of the foundations; this needs to be verified
			return { fixture: 'foundation', data: [0] };
		case 'a':
		case 'b':
		case 'c':
		case 'd':
		case 'e':
		case 'f':
			return { fixture: 'cell', data: [parseInt(p, 16) - 10] };
		default:
			throw new Error(`invalid position shorthand: "${p ?? 'undefined'}"`);
	}
}
