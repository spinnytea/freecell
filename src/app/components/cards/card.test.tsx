import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { CardImage, RankList, SuitList, units } from './card';

describe('card', () => {
	test('first', () => {
		render(<CardImage rank="5" suit="hearts" />);
		expect(screen.getByAltText('5 of hearts')).toMatchSnapshot();
	});

	describe('getFilename', () => {
		test('fancy', () => {
			const allNames: string[] = [];

			SuitList.forEach((suit) => {
				RankList.forEach((rank) => {
					allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit, true)}`);
				});
			});

			expect(allNames).toMatchSnapshot();
		});

		test('simple', () => {
			const allNames: string[] = [];

			SuitList.forEach((suit) => {
				RankList.forEach((rank) => {
					allNames.push(`${rank} of ${suit} → ${units.getFilename(rank, suit, false)}`);
				});
			});

			expect(allNames).toMatchSnapshot();
		});
	});

	describe('manual checks', () => {
		test.todo('visual check on all Suit x Rank');

		test.todo('visual check on red/black jokers');
	});
});
