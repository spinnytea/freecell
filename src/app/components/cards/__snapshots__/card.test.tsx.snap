// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card first 1`] = `
<img
  alt="5 of hearts"
  data-nimg="1"
  decoding="async"
  draggable="false"
  fetchpriority="high"
  height="242"
  src="/SVG-cards-1.3/5_of_hearts.svg"
  style="color: transparent;"
  width="167"
/>
`;

exports[`card getFilename fancy 1`] = `
[
  "ace of clubs → /SVG-cards-1.3/ace_of_clubs.svg",
  "2 of clubs → /SVG-cards-1.3/2_of_clubs.svg",
  "3 of clubs → /SVG-cards-1.3/3_of_clubs.svg",
  "4 of clubs → /SVG-cards-1.3/4_of_clubs.svg",
  "5 of clubs → /SVG-cards-1.3/5_of_clubs.svg",
  "6 of clubs → /SVG-cards-1.3/6_of_clubs.svg",
  "7 of clubs → /SVG-cards-1.3/7_of_clubs.svg",
  "8 of clubs → /SVG-cards-1.3/8_of_clubs.svg",
  "9 of clubs → /SVG-cards-1.3/9_of_clubs.svg",
  "10 of clubs → /SVG-cards-1.3/10_of_clubs.svg",
  "jack of clubs → /SVG-cards-1.3/jack_of_clubs2.svg",
  "queen of clubs → /SVG-cards-1.3/queen_of_clubs2.svg",
  "king of clubs → /SVG-cards-1.3/king_of_clubs2.svg",
  "ace of diamonds → /SVG-cards-1.3/ace_of_diamonds.svg",
  "2 of diamonds → /SVG-cards-1.3/2_of_diamonds.svg",
  "3 of diamonds → /SVG-cards-1.3/3_of_diamonds.svg",
  "4 of diamonds → /SVG-cards-1.3/4_of_diamonds.svg",
  "5 of diamonds → /SVG-cards-1.3/5_of_diamonds.svg",
  "6 of diamonds → /SVG-cards-1.3/6_of_diamonds.svg",
  "7 of diamonds → /SVG-cards-1.3/7_of_diamonds.svg",
  "8 of diamonds → /SVG-cards-1.3/8_of_diamonds.svg",
  "9 of diamonds → /SVG-cards-1.3/9_of_diamonds.svg",
  "10 of diamonds → /SVG-cards-1.3/10_of_diamonds.svg",
  "jack of diamonds → /SVG-cards-1.3/jack_of_diamonds2.svg",
  "queen of diamonds → /SVG-cards-1.3/queen_of_diamonds2.svg",
  "king of diamonds → /SVG-cards-1.3/king_of_diamonds2.svg",
  "ace of hearts → /SVG-cards-1.3/ace_of_hearts.svg",
  "2 of hearts → /SVG-cards-1.3/2_of_hearts.svg",
  "3 of hearts → /SVG-cards-1.3/3_of_hearts.svg",
  "4 of hearts → /SVG-cards-1.3/4_of_hearts.svg",
  "5 of hearts → /SVG-cards-1.3/5_of_hearts.svg",
  "6 of hearts → /SVG-cards-1.3/6_of_hearts.svg",
  "7 of hearts → /SVG-cards-1.3/7_of_hearts.svg",
  "8 of hearts → /SVG-cards-1.3/8_of_hearts.svg",
  "9 of hearts → /SVG-cards-1.3/9_of_hearts.svg",
  "10 of hearts → /SVG-cards-1.3/10_of_hearts.svg",
  "jack of hearts → /SVG-cards-1.3/jack_of_hearts2.svg",
  "queen of hearts → /SVG-cards-1.3/queen_of_hearts2.svg",
  "king of hearts → /SVG-cards-1.3/king_of_hearts2.svg",
  "ace of spades → /SVG-cards-1.3/ace_of_spades.svg",
  "2 of spades → /SVG-cards-1.3/2_of_spades.svg",
  "3 of spades → /SVG-cards-1.3/3_of_spades.svg",
  "4 of spades → /SVG-cards-1.3/4_of_spades.svg",
  "5 of spades → /SVG-cards-1.3/5_of_spades.svg",
  "6 of spades → /SVG-cards-1.3/6_of_spades.svg",
  "7 of spades → /SVG-cards-1.3/7_of_spades.svg",
  "8 of spades → /SVG-cards-1.3/8_of_spades.svg",
  "9 of spades → /SVG-cards-1.3/9_of_spades.svg",
  "10 of spades → /SVG-cards-1.3/10_of_spades.svg",
  "jack of spades → /SVG-cards-1.3/jack_of_spades2.svg",
  "queen of spades → /SVG-cards-1.3/queen_of_spades2.svg",
  "king of spades → /SVG-cards-1.3/king_of_spades2.svg",
  "joker of clubs → /SVG-cards-1.3/black_joker.svg",
  "joker of diamonds → /SVG-cards-1.3/red_joker.svg",
  "joker of hearts → /SVG-cards-1.3/red_joker.svg",
  "joker of spades → /SVG-cards-1.3/black_joker.svg",
  "card back → Card_back_10.svg",
]
`;

exports[`card getFilename simple 1`] = `
[
  "ace of clubs → /SVG-cards-1.3/ace_of_clubs.svg",
  "2 of clubs → /SVG-cards-1.3/2_of_clubs.svg",
  "3 of clubs → /SVG-cards-1.3/3_of_clubs.svg",
  "4 of clubs → /SVG-cards-1.3/4_of_clubs.svg",
  "5 of clubs → /SVG-cards-1.3/5_of_clubs.svg",
  "6 of clubs → /SVG-cards-1.3/6_of_clubs.svg",
  "7 of clubs → /SVG-cards-1.3/7_of_clubs.svg",
  "8 of clubs → /SVG-cards-1.3/8_of_clubs.svg",
  "9 of clubs → /SVG-cards-1.3/9_of_clubs.svg",
  "10 of clubs → /SVG-cards-1.3/10_of_clubs.svg",
  "jack of clubs → /SVG-cards-1.3/jack_of_clubs.svg",
  "queen of clubs → /SVG-cards-1.3/queen_of_clubs.svg",
  "king of clubs → /SVG-cards-1.3/king_of_clubs.svg",
  "ace of diamonds → /SVG-cards-1.3/ace_of_diamonds.svg",
  "2 of diamonds → /SVG-cards-1.3/2_of_diamonds.svg",
  "3 of diamonds → /SVG-cards-1.3/3_of_diamonds.svg",
  "4 of diamonds → /SVG-cards-1.3/4_of_diamonds.svg",
  "5 of diamonds → /SVG-cards-1.3/5_of_diamonds.svg",
  "6 of diamonds → /SVG-cards-1.3/6_of_diamonds.svg",
  "7 of diamonds → /SVG-cards-1.3/7_of_diamonds.svg",
  "8 of diamonds → /SVG-cards-1.3/8_of_diamonds.svg",
  "9 of diamonds → /SVG-cards-1.3/9_of_diamonds.svg",
  "10 of diamonds → /SVG-cards-1.3/10_of_diamonds.svg",
  "jack of diamonds → /SVG-cards-1.3/jack_of_diamonds.svg",
  "queen of diamonds → /SVG-cards-1.3/queen_of_diamonds.svg",
  "king of diamonds → /SVG-cards-1.3/king_of_diamonds.svg",
  "ace of hearts → /SVG-cards-1.3/ace_of_hearts.svg",
  "2 of hearts → /SVG-cards-1.3/2_of_hearts.svg",
  "3 of hearts → /SVG-cards-1.3/3_of_hearts.svg",
  "4 of hearts → /SVG-cards-1.3/4_of_hearts.svg",
  "5 of hearts → /SVG-cards-1.3/5_of_hearts.svg",
  "6 of hearts → /SVG-cards-1.3/6_of_hearts.svg",
  "7 of hearts → /SVG-cards-1.3/7_of_hearts.svg",
  "8 of hearts → /SVG-cards-1.3/8_of_hearts.svg",
  "9 of hearts → /SVG-cards-1.3/9_of_hearts.svg",
  "10 of hearts → /SVG-cards-1.3/10_of_hearts.svg",
  "jack of hearts → /SVG-cards-1.3/jack_of_hearts.svg",
  "queen of hearts → /SVG-cards-1.3/queen_of_hearts.svg",
  "king of hearts → /SVG-cards-1.3/king_of_hearts.svg",
  "ace of spades → /SVG-cards-1.3/ace_of_spades2.svg",
  "2 of spades → /SVG-cards-1.3/2_of_spades.svg",
  "3 of spades → /SVG-cards-1.3/3_of_spades.svg",
  "4 of spades → /SVG-cards-1.3/4_of_spades.svg",
  "5 of spades → /SVG-cards-1.3/5_of_spades.svg",
  "6 of spades → /SVG-cards-1.3/6_of_spades.svg",
  "7 of spades → /SVG-cards-1.3/7_of_spades.svg",
  "8 of spades → /SVG-cards-1.3/8_of_spades.svg",
  "9 of spades → /SVG-cards-1.3/9_of_spades.svg",
  "10 of spades → /SVG-cards-1.3/10_of_spades.svg",
  "jack of spades → /SVG-cards-1.3/jack_of_spades.svg",
  "queen of spades → /SVG-cards-1.3/queen_of_spades.svg",
  "king of spades → /SVG-cards-1.3/king_of_spades.svg",
  "joker of clubs → /SVG-cards-1.3/black_joker.svg",
  "joker of diamonds → /SVG-cards-1.3/red_joker.svg",
  "joker of hearts → /SVG-cards-1.3/red_joker.svg",
  "joker of spades → /SVG-cards-1.3/black_joker.svg",
  "card back → Card_back_10.svg",
]
`;
